app/src/lib/stores/helpers/pull-request-updater.ts

```diff
@@ -0,0 +1,129 @@
+const StatusInterval = 1000 * 60 * 10
+
+//** Interval to check for pull request statuses after commits have been pushed */
+const PostPushInterval = 1000 * 60
+
+enum TimeoutHandles {
+  PullRequest = 'PullRequestHandle',
+  Status = 'StatusHandle',
+  PushedPullRequest = 'PushedPullRequestHandle',
+}
+
+/**
+ * Acts as a service for downloading the latest pull request
+ * and status info from GitHub.
+ */
+export class PullRequestUpdater {
+  private readonly repository: GitHubRepository
+  private readonly account: Account
+  private readonly store: PullRequestStore
+
+  private readonly timeoutHandles = new Map<TimeoutHandles, number>()
+  private isStopped: boolean = true
+
+  private currentPullRequests: ReadonlyArray<PullRequest> = []
+
+  public constructor(
+    repository: GitHubRepository,
+    account: Account,
+    pullRequestStore: PullRequestStore
+  ) {
+    this.repository = repository
+    this.account = account
+    this.store = pullRequestStore
+  }
+
+  /** Starts the updater */
+  public start() {
+    if (!this.isStopped) {
+      fatalError(
+        'Cannot start the Pull Request Updater that is already running.'
+      )
+
+      return
+    }
+
+    this.timeoutHandles.set(
+      TimeoutHandles.PullRequest,
+
+      window.setTimeout(() => {
+        this.store.refreshPullRequests(this.repository, this.account)
+      }, PullRequestInterval)
+    )
+
+    this.timeoutHandles.set(
+      TimeoutHandles.Status,
+      window.setTimeout(() => {
+        this.store.refreshPullRequestStatuses(this.repository, this.account)
+      }, StatusInterval)
+    )
+  }
+
+  public stop() {
+    this.isStopped = true
+
+    for (const timeoutHandle of this.timeoutHandles.values()) {
+      window.clearTimeout(timeoutHandle)
+    }
+
+    this.timeoutHandles.clear()
+  }
+
+  /** Starts fetching the statuses of PRs at an accelerated rate */
+  public didPushPullRequest(pullRequest: PullRequest) {
+    if (this.currentPullRequests.find(p => p.id === pullRequest.id)) {
+      return
+    }
+
+    this.currentPullRequests = [...this.currentPullRequests, pullRequest]
+
+    const pushedPRHandle = this.timeoutHandles.get(
+      TimeoutHandles.PushedPullRequest
+    )
+    if (pushedPRHandle) {
+      return
+    }
+
+    const handle = window.setTimeout(
+      () => this.refreshPullRequestStatus(),
+      PostPushInterval
+    )
+    this.timeoutHandles.set(TimeoutHandles.PushedPullRequest, handle)
+  }
+
+  private async refreshPullRequestStatus() {
+    await this.store.refreshPullRequestStatuses(this.repository, this.account)
+    const prStatuses = await this.store.getPullRequestStatuses()
+
+    for (const prStatus of prStatuses) {
+      if (
+        !prStatus.totalCount ||
+        prStatus.state === 'success' ||
+        prStatus.state === 'failure'
+      ) {
+        this.currentPullRequests = this.currentPullRequests.filter(
+          p => p.number === prStatus.pullRequestNumber
```

Comment:
Maybe I'm missing some context here, but after updating all the pull request statuses for this repository we're:

  - iterating over the latest set of PR statuses, and for each status we are:
    - filtering the current set of pull requests to find the matching pull request for a status
    - setting `this.currentPullRequests` to this result (which will either be one entry if it matches or empty otherwise)
    - repeating this process, using the updated `this.currentPullRequests`

I was under the impression that `this.currentPullRequests` is the entire list of pull requests for the current repository - reassigning it here probably means we may never release the handle below.

So I guess I have two questions here:

 - what actual work do we need to do to `this.currentPullRequests` as part of the "you pushed a PR branch" workflow?
 - when should this cleanup and release the handle?
 



I can't see where we're using this, so my gut feeling is that this is just unnecessary work.

What's the goal for removing the handle below? Could we simplify this code and avoid the work

Reply:
Your questions uncovered a logic error! 🙇 

So the idea behind this method is to figure out when we should remove `TimeoutHandlers.PushedPullReuquests` from our list of `timeoutHandles`. We decided to use the length of `currerentPullRequests` to determine when to remove the handle, provided one exists. The logic error was that we'd set `currentPullRequests` to the last `prStatus` that wasn't pending. Specifically, this line: https://github.com/desktop/desktop/blob/cd02943bfcd0c58982bedfef280ad4cbaa2846d5/app/src/lib/stores/helpers/pull-request-updater.ts#L115
What we needed was to return all elements *except* the one that matched that condition.

So to address your questions:
> what actual work do we need to do to this.currentPullRequests as part of the "you pushed a PR branch" workflow?

1. After a push, we call `didPushPullRequest`
https://github.com/desktop/desktop/blob/cd02943bfcd0c58982bedfef280ad4cbaa2846d5/app/src/lib/stores/app-store.ts#L1818-L1825
2. We add `currentPR` to the pull request updater's local state
https://github.com/desktop/desktop/blob/cd02943bfcd0c58982bedfef280ad4cbaa2846d5/app/src/lib/stores/helpers/pull-request-updater.ts#L88
3. We set up a `timeoutHandle` that will call `refreshPullRequestStatuses` at an accelerated rate
https://github.com/desktop/desktop/blob/cd02943bfcd0c58982bedfef280ad4cbaa2846d5/app/src/lib/stores/helpers/pull-request-updater.ts#L88
4. `refreshPullRequestStatus` gets the lastest pull request statuses and removes pull requests that have a PR status of **pending** or 1 or more statuses 
https://github.com/desktop/desktop/blob/cd02943bfcd0c58982bedfef280ad4cbaa2846d5/app/src/lib/stores/helpers/pull-request-updater.ts#L108-L118
5. Remove the handle if no more PRs exists, this should always get called at some point since  every pull request in `currentPullRequest` has a status

> when should this cleanup and release the handle?

We should only clean up and release the handle if no more pull requests have a status of pending. See item 5 above.

Reply:
I'm glad to have helped! 💖 

> We should only clean up and release the handle if no more pull requests have a status of pending

Aha, this is the other little clue! Thanks for clearing it up! Could we :memo: that bit to make it clearer we're interested in `pending`, and maybe make the `if` statement a bit more concise (it should be a string enum) ?

```ts
if (!prStatus.TotalCount || prStatus.state !== 'pending') {
  // update currentPullRequests
}
```